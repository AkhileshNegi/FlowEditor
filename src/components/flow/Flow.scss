@import '~/variables.scss';

.loading {
    visibility: hidden;
    height: inherit;
}

.loaded {
    height: inherit;
    position: relative;
}

.nodeList {
    position: relative;
    user-select: none;
}

.dragSelection {
    position: absolute;
    background: red;
    border: 0px solid darken($selection_square, 10%);
    opacity: 0;
    pointer-events: none;
    z-index: $z_selection;
}
