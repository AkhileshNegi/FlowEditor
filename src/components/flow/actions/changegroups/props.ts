import { ActionFormHelper, Type } from '~/config/typeConfigs';
import { ChangeGroups } from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export interface ChangeGroupsFormProps {
    // action details
    nodeSettings: NodeEditorSettings;
    formHelper: ActionFormHelper;
    typeConfig: Type;

    // update handlers
    updateAction(action: ChangeGroups): void;

    // modal notifiers
    onTypeChange(config: Type): void;
    onClose(canceled: boolean): void;
}
export default ChangeGroupsFormProps;
