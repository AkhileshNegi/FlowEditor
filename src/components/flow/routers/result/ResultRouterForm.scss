@import '~/variables.scss';

.lead-in {
    margin-bottom: 5px;
    display: inline-block;
    vertical-align: top;
    margin-top: 6px;
}

.lead-in-sub {
    margin-bottom: 5px;
    display: inline-block;
    vertical-align: top;
    margin-top: 6px;
    margin-left: 8px;
}

.result-select {
    width: 254px;
    display: inline-block;
    margin-left: 10px;
    margin-bottom: -10px;
    vertical-align: top;
}

.result-select-delimited {
    width: 166px;
    display: inline-block;
    margin-left: 10px;
    margin-bottom: -10px;
    vertical-align: top;
}

.should-delimit {
    margin-top: 20px;
}

.field-number {
    width: 96px;
    display: inline-block;
    margin-left: 10px;
    margin-bottom: -10px;
    vertical-align: top;
}

.delimiter {
    width: 158px;
    display: inline-block;
    margin-left: 12px;
    margin-bottom: -10px;
    vertical-align: top;
}
