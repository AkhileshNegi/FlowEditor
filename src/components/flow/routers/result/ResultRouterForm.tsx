import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet, HeaderStyle, Tab } from '~/components/dialog/Dialog';
import { RouterFormProps } from '~/components/flow/props';
import CaseList, { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import AssetSelector from '~/components/form/assetselector/AssetSelector';
import CheckboxElement from '~/components/form/checkbox/CheckboxElement';
import OptionalTextInput from '~/components/form/optionaltext/OptionalTextInput';
import SelectElement, { SelectOption } from '~/components/form/select/SelectElement';
import TypeList from '~/components/nodeeditor/TypeList';
import { fakePropType } from '~/config/ConfigProvider';
import { Asset } from '~/store/flowContext';
import { AssetEntry, FormState, mergeForm, StringEntry } from '~/store/nodeEditor';
import { validate, validateRequired } from '~/store/validators';
import { small } from '~/utils/reactselect';

import {
    DELIMITER_OPTIONS,
    FIELD_NUMBER_OPTIONS,
    getDelimiterOption,
    getFieldOption,
    nodeToState,
    stateToNode
} from './helpers';
import * as styles from './ResultRouterForm.scss';

export interface ResultRouterFormState extends FormState {
    result: AssetEntry;
    cases: CaseProps[];
    resultName: StringEntry;
    shouldDelimit: boolean;

    fieldNumber: number;
    delimiter: string;
}

export const leadInSpecId = 'lead-in';

export default class ResultRouterForm extends React.Component<
    RouterFormProps,
    ResultRouterFormState
> {
    constructor(props: RouterFormProps) {
        super(props);

        this.state = nodeToState(this.props.nodeSettings);

        bindCallbacks(this, {
            include: [/^on/, /^handle/]
        });
    }

    public static contextTypes = {
        assetService: fakePropType
    };

    private handleUpdateResultName(value: string): void {
        this.setState({ resultName: { value } });
    }

    private handleResultChanged(selected: Asset[]): boolean {
        const updates: Partial<ResultRouterFormState> = {
            result: validate('Result to split on', selected[0], [validateRequired])
        };

        const updated = mergeForm(this.state, updates);
        this.setState(updated);
        return updated.valid;
    }

    private handleCasesUpdated(cases: CaseProps[]): void {
        this.setState({ cases });
    }

    private handleSave(): void {
        const valid = this.handleResultChanged([this.state.result.value]);
        if (valid) {
            this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
            this.props.onClose(false);
        }
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave, disabled: !this.state.valid },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    private handleShouldDelimitChanged(checked: boolean): void {
        this.setState({ shouldDelimit: checked });
    }

    private handleFieldNumberChanged(selected: SelectOption): void {
        this.setState({ fieldNumber: parseInt(selected.value, 10) });
    }

    private handleDelimiterChanged(selected: SelectOption): void {
        this.setState({ delimiter: selected.value });
    }

    private renderField(): JSX.Element {
        return (
            <>
                <div className={styles.leadIn}>If the flow result</div>
                <div className={styles.resultSelect}>
                    <AssetSelector
                        entry={this.state.result}
                        styles={small}
                        name="Flow Result"
                        placeholder="Select Result"
                        searchable={true}
                        assets={this.props.assetStore.results}
                        onChange={this.handleResultChanged}
                    />
                </div>
            </>
        );
    }

    private renderFieldDelimited(): JSX.Element {
        return (
            <>
                <div className={styles.leadIn}>If the</div>
                <div className={styles.fieldNumber}>
                    <SelectElement
                        styles={small}
                        name="Field Number"
                        entry={{ value: getFieldOption(this.state.fieldNumber) }}
                        onChange={this.handleFieldNumberChanged}
                        options={FIELD_NUMBER_OPTIONS}
                    />
                </div>
                <div className={styles.leadInSub}>field of</div>
                <div className={styles.resultSelectDelimited}>
                    <AssetSelector
                        entry={this.state.result}
                        styles={small}
                        name="Flow Result"
                        placeholder="Select Result"
                        searchable={true}
                        assets={this.props.assetStore.results}
                        onChange={this.handleResultChanged}
                    />
                </div>
                <div className={styles.leadInSub}>delimited by</div>
                <div className={styles.delimiter}>
                    <SelectElement
                        styles={small}
                        name="Delimiter"
                        entry={{ value: getDelimiterOption(this.state.delimiter) }}
                        onChange={this.handleDelimiterChanged}
                        options={DELIMITER_OPTIONS}
                    />
                </div>
            </>
        );
    }

    public render(): JSX.Element {
        const typeConfig = this.props.typeConfig;

        const back = (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                headerStyle={HeaderStyle.BARBER}
                subtitle="Advanced Settings"
                buttons={this.getButtons()}
                headerIcon="fe-cog"
            />
        );

        const advanced: Tab = {
            name: 'Advanced',
            body: (
                <div className={styles.shouldDelimit}>
                    <CheckboxElement
                        name="Delimit"
                        title="Delimit Result"
                        checked={this.state.shouldDelimit}
                        description="Evaluate your rules against a delimited part of your result"
                        onChange={this.handleShouldDelimitChanged}
                    />
                </div>
            ),
            checked: this.state.shouldDelimit
        };

        return (
            <Dialog
                title={typeConfig.name}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
                tabs={[advanced]}
            >
                <TypeList
                    __className=""
                    initialType={typeConfig}
                    onChange={this.props.onTypeChange}
                />

                {this.state.shouldDelimit ? this.renderFieldDelimited() : this.renderField()}

                <CaseList
                    data-spec="cases"
                    cases={this.state.cases}
                    onCasesUpdated={this.handleCasesUpdated}
                />
                <OptionalTextInput
                    name="Result Name"
                    value={this.state.resultName}
                    onChange={this.handleUpdateResultName}
                    toggleText="Save as.."
                    helpText="By naming the result, you can reference it later using @run.results.whatever_the_name_is"
                />
            </Dialog>
        );
    }
}
