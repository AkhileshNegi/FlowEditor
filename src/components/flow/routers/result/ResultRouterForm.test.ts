import { RouterFormProps } from "components/flow/props";
import ResultRouterForm from "components/flow/routers/result/ResultRouterForm";
import { Types } from "config/interfaces";
import { AssetType } from "store/flowContext";
import { composeComponentTestUtils, mock } from "testUtils";
import {
  createRenderNode,
  getRouterFormProps,
  createMatchRouter
} from "testUtils/assetCreators";
import * as utils from "utils";

const routerNode = createMatchRouter([]);
routerNode.ui = {
  position: { left: 0, top: 0 },
  type: Types.split_by_run_result,
  config: {
    id: "favorite_color",
    type: AssetType.Result
  }
};

const { setup } = composeComponentTestUtils<RouterFormProps>(
  ResultRouterForm,
  getRouterFormProps(routerNode)
);

mock(utils, "createUUID", utils.seededUUIDs());

describe(ResultRouterForm.name, () => {
  it("should render", () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });
});
