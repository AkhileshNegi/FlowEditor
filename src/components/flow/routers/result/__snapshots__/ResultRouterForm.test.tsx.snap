// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultRouterForm should a fielded operand if configured 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "router": Object {
        "cases": Array [],
        "categories": Array [
          Object {
            "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            "name": "All Responses",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
        ],
        "default_category_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        "operand": "@(field(results.my_test_result, 0, \\"+\\"))",
        "type": "switch",
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
    },
    "ui": Object {
      "config": Object {
        "cases": Object {},
        "delimiter": "+",
        "index": 0,
        "operand": Object {
          "id": "my_test_result",
          "name": "My Test Result",
          "type": "result",
        },
      },
      "position": null,
      "type": "split_by_run_result_delimited",
    },
  },
]
`;

exports[`ResultRouterForm should create the right operand on save 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "router": Object {
        "cases": Array [],
        "categories": Array [
          Object {
            "exit_uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
            "name": "All Responses",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
        ],
        "default_category_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        "operand": "@results.my_test_result",
        "type": "switch",
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
    },
    "ui": Object {
      "config": Object {
        "cases": Object {},
        "operand": Object {
          "id": "my_test_result",
          "name": "My Test Result",
          "type": "result",
        },
      },
      "position": null,
      "type": "split_by_run_result",
    },
  },
]
`;

exports[`ResultRouterForm should render 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="tabs"
      >
        <div
          class="tab "
        >
          Advanced
           
        </div>
      </div>
      <div
        class="header split_by_run_result"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split by Flow Result
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                name="type"
                options="[{\\"name\\":\\"Send the contact a message\\",\\"value\\":\\"send_msg\\"},{\\"name\\":\\"Wait for the contact to respond\\",\\"value\\":\\"wait_for_response\\"},{\\"name\\":\\"Send somebody else a message\\",\\"value\\":\\"send_broadcast\\"},{\\"name\\":\\"Label the incoming message\\",\\"value\\":\\"add_input_labels\\"},{\\"name\\":\\"Add a URN for the contact\\",\\"value\\":\\"add_contact_urn\\"},{\\"name\\":\\"Add the contact to a group\\",\\"value\\":\\"add_contact_groups\\"},{\\"name\\":\\"Remove the contact from a group\\",\\"value\\":\\"remove_contact_groups\\"},{\\"name\\":\\"Update the contact\\",\\"value\\":\\"set_contact_field\\"},{\\"name\\":\\"Send an email\\",\\"value\\":\\"send_email\\"},{\\"name\\":\\"Save a result for this flow\\",\\"value\\":\\"set_run_result\\"},{\\"name\\":\\"Call a webhook\\",\\"value\\":\\"call_webhook\\"},{\\"name\\":\\"Call Zapier\\",\\"value\\":\\"call_resthook\\"},{\\"name\\":\\"Enter another flow\\",\\"value\\":\\"enter_flow\\"},{\\"name\\":\\"Start somebody else in a flow\\",\\"value\\":\\"start_session\\"},{\\"name\\":\\"Open a ticket with a human agent\\",\\"value\\":\\"open_ticket\\"},{\\"name\\":\\"Send the contact airtime\\",\\"value\\":\\"transfer_airtime\\"},{\\"name\\":\\"Split by intent\\",\\"value\\":\\"split_by_intent\\"},{\\"name\\":\\"Split by a custom expression\\",\\"value\\":\\"split_by_expression\\"},{\\"name\\":\\"Split by a contact field\\",\\"value\\":\\"split_by_contact_field\\"},{\\"name\\":\\"Split by a result in the flow\\",\\"value\\":\\"split_by_run_result\\"},{\\"name\\":\\"Split by random chance\\",\\"value\\":\\"split_by_random\\"},{\\"name\\":\\"Split by group membership\\",\\"value\\":\\"split_by_groups\\"},{\\"name\\":\\"Split by URN type\\",\\"value\\":\\"split_by_scheme\\"}]"
                searchable="true"
                values="[{\\"name\\":\\"Split by a result in the flow\\",\\"value\\":\\"split_by_run_result\\"}]"
              />
            </div>
          </div>
        </div>
        <div
          class="non_delimited"
        >
          <div
            class="lead_in"
          >
            If the flow result
          </div>
          <div
            class="result_select"
          >
            <div
              class="ele"
            >
              <div
                class="react-select selection css-2b097c-container"
              >
                <div
                  class=" css-p773gl-control"
                >
                  <div
                    class=" css-c2yhe2"
                  >
                    <div
                      class=" css-1wa3eu0-placeholder"
                    >
                      Select Result
                    </div>
                    <input
                      class="css-62g3xt-dummyInput"
                      id="react-select-2-input"
                      readonly=""
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class=" css-1ywtfjx-IndicatorsContainer"
                  >
                    <span
                      class=" css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="case_list_container "
        >
          <div
            class="case_list"
          >
            <div
              class="kase case_list_case"
            >
              <div
                class="ele group"
              >
                <div
                  class="kase has_any_word"
                  data-draggable="true"
                >
                  <span
                    class="fe-chevrons-expand dnd_icon"
                    data-draggable="true"
                  />
                  <div
                    class="choice"
                  >
                    <div
                      class="small "
                    >
                      <temba-select
                        data-testid="temba_select_operator"
                        name="operator"
                        options="[{\\"name\\":\\"has any of the words\\",\\"value\\":\\"has_any_word\\"},{\\"name\\":\\"has all of the words\\",\\"value\\":\\"has_all_words\\"},{\\"name\\":\\"has the phrase\\",\\"value\\":\\"has_phrase\\"},{\\"name\\":\\"has only the phrase\\",\\"value\\":\\"has_only_phrase\\"},{\\"name\\":\\"starts with\\",\\"value\\":\\"has_beginning\\"},{\\"name\\":\\"has some text\\",\\"value\\":\\"has_text\\"},{\\"name\\":\\"has a number\\",\\"value\\":\\"has_number\\"},{\\"name\\":\\"has a number between\\",\\"value\\":\\"has_number_between\\"},{\\"name\\":\\"has a number below\\",\\"value\\":\\"has_number_lt\\"},{\\"name\\":\\"has a number at or below\\",\\"value\\":\\"has_number_lte\\"},{\\"name\\":\\"has a number equal to\\",\\"value\\":\\"has_number_eq\\"},{\\"name\\":\\"has a number at or above\\",\\"value\\":\\"has_number_gte\\"},{\\"name\\":\\"has a number above\\",\\"value\\":\\"has_number_gt\\"},{\\"name\\":\\"has a date\\",\\"value\\":\\"has_date\\"},{\\"name\\":\\"has a date before\\",\\"value\\":\\"has_date_lt\\"},{\\"name\\":\\"has a date equal to\\",\\"value\\":\\"has_date_eq\\"},{\\"name\\":\\"has a date after\\",\\"value\\":\\"has_date_gt\\"},{\\"name\\":\\"has a time\\",\\"value\\":\\"has_time\\"},{\\"name\\":\\"has a phone number\\",\\"value\\":\\"has_phone\\"},{\\"name\\":\\"has an email\\",\\"value\\":\\"has_email\\"},{\\"name\\":\\"has state\\",\\"value\\":\\"has_state\\"},{\\"name\\":\\"has district\\",\\"value\\":\\"has_district\\"},{\\"name\\":\\"has ward\\",\\"value\\":\\"has_ward\\"},{\\"name\\":\\"matches regex\\",\\"value\\":\\"has_pattern\\"}]"
                        values="[{\\"name\\":\\"has any of the words\\",\\"value\\":\\"has_any_word\\"}]"
                      />
                    </div>
                  </div>
                  <div
                    class="single_operand"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper"
                      >
                        <input
                          class="textinput"
                          data-spec="input"
                          data-testid="input"
                          maxlength="-1"
                          name="arguments"
                          placeholder=""
                          type="text"
                          value=""
                        />
                        <div
                          class="completion_container hidden"
                          data-spec="completion-options"
                          style="top: 0px; left: 0px;"
                        >
                          <div
                            class="options_wrapper"
                          >
                            <ul
                              class="option_list"
                              data-spec="completion-list"
                            />
                            <div
                              class="help"
                              data-spec="completion-help"
                            >
                              Tab to complete, enter to select
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="categorize_as"
                    data-draggable="true"
                  >
                    categorize as
                  </div>
                  <div
                    class="category"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper"
                      >
                        <input
                          class="textinput"
                          data-spec="input"
                          data-testid="input"
                          maxlength="36"
                          name="Exit Name"
                          type="text"
                          value=""
                        />
                        <div
                          class="completion_container hidden"
                          data-spec="completion-options"
                          style="top: 0px; left: 0px;"
                        >
                          <div
                            class="options_wrapper"
                          >
                            <ul
                              class="option_list"
                              data-spec="completion-list"
                            />
                            <div
                              class="help"
                              data-spec="completion-help"
                            >
                              Tab to complete, enter to select
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <span
                    class="fe-x remove_icon"
                    data-testid="remove-case-b1f332f3-bdd3-4891-aec5-1843a712dbf1"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`ResultRouterForm should show delimit options 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="tabs"
      >
        <div
          class="tab "
        >
          Advanced
           
          <span
            class="tab_icon fe-check"
          />
        </div>
      </div>
      <div
        class="header split_by_run_result"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split by Flow Result
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                name="type"
                options="[{\\"name\\":\\"Send the contact a message\\",\\"value\\":\\"send_msg\\"},{\\"name\\":\\"Wait for the contact to respond\\",\\"value\\":\\"wait_for_response\\"},{\\"name\\":\\"Send somebody else a message\\",\\"value\\":\\"send_broadcast\\"},{\\"name\\":\\"Label the incoming message\\",\\"value\\":\\"add_input_labels\\"},{\\"name\\":\\"Add a URN for the contact\\",\\"value\\":\\"add_contact_urn\\"},{\\"name\\":\\"Add the contact to a group\\",\\"value\\":\\"add_contact_groups\\"},{\\"name\\":\\"Remove the contact from a group\\",\\"value\\":\\"remove_contact_groups\\"},{\\"name\\":\\"Update the contact\\",\\"value\\":\\"set_contact_field\\"},{\\"name\\":\\"Send an email\\",\\"value\\":\\"send_email\\"},{\\"name\\":\\"Save a result for this flow\\",\\"value\\":\\"set_run_result\\"},{\\"name\\":\\"Call a webhook\\",\\"value\\":\\"call_webhook\\"},{\\"name\\":\\"Call Zapier\\",\\"value\\":\\"call_resthook\\"},{\\"name\\":\\"Enter another flow\\",\\"value\\":\\"enter_flow\\"},{\\"name\\":\\"Start somebody else in a flow\\",\\"value\\":\\"start_session\\"},{\\"name\\":\\"Open a ticket with a human agent\\",\\"value\\":\\"open_ticket\\"},{\\"name\\":\\"Send the contact airtime\\",\\"value\\":\\"transfer_airtime\\"},{\\"name\\":\\"Split by intent\\",\\"value\\":\\"split_by_intent\\"},{\\"name\\":\\"Split by a custom expression\\",\\"value\\":\\"split_by_expression\\"},{\\"name\\":\\"Split by a contact field\\",\\"value\\":\\"split_by_contact_field\\"},{\\"name\\":\\"Split by a result in the flow\\",\\"value\\":\\"split_by_run_result\\"},{\\"name\\":\\"Split by random chance\\",\\"value\\":\\"split_by_random\\"},{\\"name\\":\\"Split by group membership\\",\\"value\\":\\"split_by_groups\\"},{\\"name\\":\\"Split by URN type\\",\\"value\\":\\"split_by_scheme\\"}]"
                searchable="true"
                values="[{\\"name\\":\\"Split by a result in the flow\\",\\"value\\":\\"split_by_run_result\\"}]"
              />
            </div>
          </div>
        </div>
        <div
          class="delimited"
        >
          <div
            class="lead_in"
          >
            If the
          </div>
          <div
            class="field_number"
          >
            <div
              class="ele"
            >
              <div
                class="small "
              >
                <temba-select
                  data-testid="temba_select_field_number"
                  name="Field Number"
                  options="[{\\"name\\":\\"first\\",\\"value\\":\\"0\\"},{\\"name\\":\\"second\\",\\"value\\":\\"1\\"},{\\"name\\":\\"third\\",\\"value\\":\\"2\\"},{\\"name\\":\\"fourth\\",\\"value\\":\\"3\\"},{\\"name\\":\\"fifth\\",\\"value\\":\\"4\\"},{\\"name\\":\\"sixth\\",\\"value\\":\\"5\\"},{\\"name\\":\\"seventh\\",\\"value\\":\\"6\\"},{\\"name\\":\\"eighth\\",\\"value\\":\\"7\\"},{\\"name\\":\\"ninth\\",\\"value\\":\\"8\\"},{\\"name\\":\\"tenth\\",\\"value\\":\\"9\\"},{\\"name\\":\\"11th\\",\\"value\\":\\"10\\"},{\\"name\\":\\"12th\\",\\"value\\":\\"11\\"},{\\"name\\":\\"13th\\",\\"value\\":\\"12\\"},{\\"name\\":\\"14th\\",\\"value\\":\\"13\\"},{\\"name\\":\\"15th\\",\\"value\\":\\"14\\"},{\\"name\\":\\"16th\\",\\"value\\":\\"15\\"},{\\"name\\":\\"17th\\",\\"value\\":\\"16\\"},{\\"name\\":\\"18th\\",\\"value\\":\\"17\\"},{\\"name\\":\\"19th\\",\\"value\\":\\"18\\"},{\\"name\\":\\"20th\\",\\"value\\":\\"19\\"}]"
                  values="[{\\"name\\":\\"second\\",\\"value\\":\\"1\\"}]"
                />
              </div>
            </div>
          </div>
          <div
            class="lead_in_sub"
          >
            field of
          </div>
          <div
            class="result_select_delimited"
          >
            <div
              class="ele"
            >
              <div
                class="react-select selection css-2b097c-container"
              >
                <div
                  class=" css-p773gl-control"
                >
                  <div
                    class=" css-c2yhe2"
                  >
                    <div
                      class=" css-1wa3eu0-placeholder"
                    >
                      Select Result
                    </div>
                    <input
                      class="css-62g3xt-dummyInput"
                      id="react-select-4-input"
                      readonly=""
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class=" css-1ywtfjx-IndicatorsContainer"
                  >
                    <span
                      class=" css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="lead_in_sub"
          >
            delimited by
          </div>
          <div
            class="delimiter"
          >
            <div
              class="ele"
            >
              <div
                class="small "
              >
                <temba-select
                  data-testid="temba_select_delimiter"
                  name="Delimiter"
                  options="[{\\"name\\":\\"spaces\\",\\"value\\":\\" \\"},{\\"name\\":\\"periods\\",\\"value\\":\\".\\"},{\\"name\\":\\"plusses\\",\\"value\\":\\"+\\"}]"
                  values="[{\\"name\\":\\"spaces\\",\\"value\\":\\" \\"}]"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="case_list_container "
        >
          <div
            class="case_list"
          >
            <div
              class="kase case_list_case"
            >
              <div
                class="ele group"
              >
                <div
                  class="kase has_any_word"
                  data-draggable="true"
                >
                  <span
                    class="fe-chevrons-expand dnd_icon"
                    data-draggable="true"
                  />
                  <div
                    class="choice"
                  >
                    <div
                      class="small "
                    >
                      <temba-select
                        data-testid="temba_select_operator"
                        name="operator"
                        options="[{\\"name\\":\\"has any of the words\\",\\"value\\":\\"has_any_word\\"},{\\"name\\":\\"has all of the words\\",\\"value\\":\\"has_all_words\\"},{\\"name\\":\\"has the phrase\\",\\"value\\":\\"has_phrase\\"},{\\"name\\":\\"has only the phrase\\",\\"value\\":\\"has_only_phrase\\"},{\\"name\\":\\"starts with\\",\\"value\\":\\"has_beginning\\"},{\\"name\\":\\"has some text\\",\\"value\\":\\"has_text\\"},{\\"name\\":\\"has a number\\",\\"value\\":\\"has_number\\"},{\\"name\\":\\"has a number between\\",\\"value\\":\\"has_number_between\\"},{\\"name\\":\\"has a number below\\",\\"value\\":\\"has_number_lt\\"},{\\"name\\":\\"has a number at or below\\",\\"value\\":\\"has_number_lte\\"},{\\"name\\":\\"has a number equal to\\",\\"value\\":\\"has_number_eq\\"},{\\"name\\":\\"has a number at or above\\",\\"value\\":\\"has_number_gte\\"},{\\"name\\":\\"has a number above\\",\\"value\\":\\"has_number_gt\\"},{\\"name\\":\\"has a date\\",\\"value\\":\\"has_date\\"},{\\"name\\":\\"has a date before\\",\\"value\\":\\"has_date_lt\\"},{\\"name\\":\\"has a date equal to\\",\\"value\\":\\"has_date_eq\\"},{\\"name\\":\\"has a date after\\",\\"value\\":\\"has_date_gt\\"},{\\"name\\":\\"has a time\\",\\"value\\":\\"has_time\\"},{\\"name\\":\\"has a phone number\\",\\"value\\":\\"has_phone\\"},{\\"name\\":\\"has an email\\",\\"value\\":\\"has_email\\"},{\\"name\\":\\"has state\\",\\"value\\":\\"has_state\\"},{\\"name\\":\\"has district\\",\\"value\\":\\"has_district\\"},{\\"name\\":\\"has ward\\",\\"value\\":\\"has_ward\\"},{\\"name\\":\\"matches regex\\",\\"value\\":\\"has_pattern\\"}]"
                        values="[{\\"name\\":\\"has any of the words\\",\\"value\\":\\"has_any_word\\"}]"
                      />
                    </div>
                  </div>
                  <div
                    class="single_operand"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper"
                      >
                        <input
                          class="textinput"
                          data-spec="input"
                          data-testid="input"
                          maxlength="-1"
                          name="arguments"
                          placeholder=""
                          type="text"
                          value=""
                        />
                        <div
                          class="completion_container hidden"
                          data-spec="completion-options"
                          style="top: 0px; left: 0px;"
                        >
                          <div
                            class="options_wrapper"
                          >
                            <ul
                              class="option_list"
                              data-spec="completion-list"
                            />
                            <div
                              class="help"
                              data-spec="completion-help"
                            >
                              Tab to complete, enter to select
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="categorize_as"
                    data-draggable="true"
                  >
                    categorize as
                  </div>
                  <div
                    class="category"
                  >
                    <div
                      class="ele"
                    >
                      <div
                        class="wrapper"
                      >
                        <input
                          class="textinput"
                          data-spec="input"
                          data-testid="input"
                          maxlength="36"
                          name="Exit Name"
                          type="text"
                          value=""
                        />
                        <div
                          class="completion_container hidden"
                          data-spec="completion-options"
                          style="top: 0px; left: 0px;"
                        >
                          <div
                            class="options_wrapper"
                          >
                            <ul
                              class="option_list"
                              data-spec="completion-list"
                            />
                            <div
                              class="help"
                              data-spec="completion-help"
                            >
                              Tab to complete, enter to select
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                  <span
                    class="fe-x remove_icon"
                    data-testid="remove-case-061fc171-8b79-4636-b892-bd0ea5aa9b42"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
